README.md
# Nintendo Wishlist Component

[![hacs_badge](https://img.shields.io/badge/HACS-Default-orange.svg)](https://github.com/custom-components/hacs)

## NOTE: This component currently only works in certain countries.  See [Supported Countries](#supported-countries) below.

Home Assistant component that keeps track of Nintendo Switch games that are on
sale on your wish list in home assistant. (There is currently not a way to get your
wish list from Nintendo, so you have to keep track of it in this component.)

[![nitendo wishlist card](https://raw.githubusercontent.com/custom-cards/nintendo-wishlist-card/master/cover-art.png)](https://raw.githubusercontent.com/custom-cards/nintendo-wishlist-card/master/cover-art.png)

## HACS Installation

1. Search for `Nintendo Wishlist Component` in the HACS Store tab.
2. Add the code to your `configuration.yaml` using the config options below.
3. **You will need to restart after installation for the component to start working.**
4. Display it in your lovelace UI using the [nintendo-wishlist-card](https://github.com/custom-cards/nintendo-wishlist-card).

## Platform Configuration

|Name|Required|Description|
|-|-|-|
|country|yes|The 2 letter country code.  See [Supported Countries](#supported-countries) below.|
|whishlist|yes|A list of Nintendo Switch titles|

## Sample Sensor Configuration

*Note:* If the title contains characters that might confuse yaml like a `:` or `'` quote the entire string
like the last game in the example below.

    sensor:
    - platform: nintendo_wishlist
      country: US
      wishlist:
        - Katana ZERO
        - OKAMI HD
        - Salt and Sanctuary
        - Dead Cells
        - Bloodstained
        - Dark Souls
        - Velocity X
        - "The Legend of Zelda: Breath of the Wild"

### Supported Countries

|Country Code|Country Name|
|-|-|
|AT|Austria|
|BE|Belgium|
|CA|Canada|
|CH|Schweiz/Suisse/Svizzera|
|DE|Germany|
|ES|Spain|
|FR|France|
|GB|UK/Ireland|
|IT|Italy|
|NL|Netherlands|
|PT|Portugal|
|RU|Russia|
|US|United States|
|ZA|South Africa|


### How wish list matching works

Currently wish list matching uses a very simple string comparison.  It's
possible to get false positives as it checks to see if any game on your wish
list starts with any title that is on sale.  In order to avoid false positives
try to have the title match that on the nintendo e-store as closely as possible.

Conversely, if you would like to match multiple titles that have a similar name
you can specify less of the title.  An example of this would be to add an item
to your wishlist like `Shantae`.  This would match any of the Shantae titles on
the e-shop (e.g. `Shantae and the Pirate's Curse` and `Shantae: Half-Genie Hero`.

### Sensors for Automations

When the custom component is run it will create a new sensor named `sensor.nintendo_wishlist`.
It's state will be the total number of games from your wish list that are on sale.
This sensor can be used with the [custom card](https://github.com/custom-cards/nintendo-wishlist-card).

The component will additionally create a sensor per title in your wish list.  These
will be named `sensor.nintendo_wishlist_{title_name}`.  For example Mega Man 11
would be `sensor.nintendo_wishlist_mega_man_11`.  The state of each sensor will
be `0` if it is not on sale and `1` if it is on sale.  You can use these sensors
in your automations to send notifications, blink your lights, or any other
automation you would like to do when a title on your wish list goes on sale.

[![example sensors](https://raw.githubusercontent.com/custom-components/sensor.nintendo_wishlist/master/sensors.png)](https://raw.githubusercontent.com/custom-components/sensor.nintendo_wishlist/master/sensors.png)

#### Example NodeRed Flow

You can import an [example of a node-red flow](https://raw.githubusercontent.com/custom-components/sensor.nintendo_wishlist/master/example_node_red_flow.json) that sends an alert to kodi when
a game on your wish list is on sale.  You can use the inject node to test the
flow and customize it to your preferences.

[![example nodered flow](https://raw.githubusercontent.com/custom-components/sensor.nintendo_wishlist/master/example_node_red_flow.png)](https://raw.githubusercontent.com/custom-components/sensor.nintendo_wishlist/master/example_node_red_flow.png)


CHANGELOG.md
# CHANGELOG

## v2.2.0 (2019-12-24)

* Updated to parse all games on sale.  Previously it was only looking at
  the first 350 results so it's possible it would miss finding a game on your
  wish list.

## v2.1.1 (2019-08-12)

* Updated the friendly name for each created sensor to just be the game title.

## v2.1.0 (2019-08-04)

* Added a sensor for each game in the wish list in addition to the previous
  sensor that contained all games on the wish list.

## v2.0.3 (2019-08-01)

* Increase the number of search results to 350 as we weren't looking at all
  games on sale (only the first 250 and there were 306 results).

## v2.0.2 (2019-07-31)

* Fixed bug that caused an error for certain countries (Schweiz, Ireland, UK,
  and Belgium).

## v2.0.1 (2019-07-30)

* Fixed bug where non North American countries would show the `$` in the price
  along with the local currency.

## v2.0.0 (2019-07-23)

* `country` is now required in the sensor platform configuration.
* Now supports 12 other countries.


custom_components/nintendo_wishlist/sensor.py
import enum
import logging
import math
from copy import copy
from datetime import timedelta
from typing import Any, Dict, List

from algoliasearch.search_client import SearchClient
import voluptuous as vol

import homeassistant.helpers.config_validation as cv
from homeassistant.components.sensor import PLATFORM_SCHEMA
from homeassistant.helpers.entity import Entity
from homeassistant.util import slugify


__version__ = "2.2.0"
DOMAIN = "nintendo_wishlist"

DEFAULT_NAME = "Nintendo Wishlist Sensor"

SCAN_INTERVAL = timedelta(minutes=10)

# North American countries.
NA_COUNTRIES = ("CA", "US")
NA_INDEX_NAMES: Dict[str, str] = {
    "CA": "noa_aem_game_en_ca",
    "US": "noa_aem_game_en_us",
}

# Mapping of country code to language code.  NOTE: language must be lowercase
# in the URL to work.
COUNTRY_LANG: Dict[str, str] = {
    "AT": "at",
    "BE": "nl",
    "CA": "en",
    "CH": "de",
    "DE": "de",
    "ES": "es",
    "FR": "fr",
    "GB": "en",
    "IT": "it",
    "NL": "nl",
    "PT": "pt",
    "RU": "ru",
    "US": "en",
    "ZA": "za",
}


class Country(enum.Enum):
    """Enum for allowed countries."""

    AT = "Austria"
    BE = "Belgium"
    CA = "Canada"
    CH = "Schweiz/Suisse/Svizzera"
    DE = "Germany"
    ES = "Spain"
    FR = "France"
    GB = "UK/Ireland"
    IT = "Italy"
    NL = "Netherlands"
    PT = "Portugal"
    RU = "Russia"
    US = "United States"
    ZA = "South Africa"


CONF_COUNTRY = "country"
CONF_WISHLIST = "wishlist"
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_WISHLIST): cv.ensure_list,
        vol.Required(CONF_COUNTRY): cv.enum(Country),
    }
)

# These are specific to European countries.
EU_SEARCH_URL = "https://searching.nintendo-europe.com/{language}/select?q=*&fq=type%3AGAME%20AND%20((playable_on_txt%3A%22HAC%22)%20AND%20(price_has_discount_b%3A%22true%22))%20AND%20sorting_title%3A*%20AND%20*%3A*&sort=price_discount_percentage_f%20desc%2C%20price_lowest_f%20desc&start=0&rows=500&wt=json&bf=linear(ms(priority%2CNOW%2FHOUR)%2C1.1e-11%2C0)"  # noqa
EU_PRICE_URL = "https://api.ec.nintendo.com/v1/price"

# Below constants used by North America (US and CA)
APP_ID = "U3B6GR4UA3"
API_KEY = "9a20c93440cf63cf1a7008d75f7438bf"
QUERIES = [
    {
        "indexName": "noa_aem_game_en_us",
        "params": "query=&hitsPerPage=350&maxValuesPerFacet=30&facets=%5B%22generalFilters%22%2C%22platform%22%2C%22availability%22%2C%22categories%22%2C%22filterShops%22%2C%22virtualConsole%22%2C%22characters%22%2C%22priceRange%22%2C%22esrb%22%2C%22filterPlayers%22%5D&tagFilters=&facetFilters=%5B%5B%22generalFilters%3ADeals%22%5D%2C%5B%22platform%3ANintendo%20Switch%22%5D%5D",  # noqa
    },
]
_LOGGER = logging.getLogger(__name__)


async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
    """Setup the sensor platform."""
    sensors = [NintendoWishlistSensor(config, game) for game in config["wishlist"]]
    sensors.append(NintendoWishlistSensor(config))
    async_add_entities(sensors, True)


def get_percent_off(original_price: float, sale_price: float) -> int:
    """Returns the percentage off of the sale price vs original price.

    We round up and return an int.

    :param original_price: The original price.
    :param sale_price: The sale price.
    :returns: The percentage as an int.
    """
    return 100 - math.ceil(100 / (original_price / sale_price))


class NintendoWishlistSensor(Entity):
    """Representation of a sensor."""

    def __init__(self, config, game: str = None):
        import aiohttp

        self.attrs = {}
        self.country = config["country"].name
        self.wishlist = [g.lower() for g in config["wishlist"]]
        self.session = aiohttp.ClientSession()
        self.game = None
        # This attribute holds the title before we lowercase it.
        self._game = None
        if game is not None:
            self._game = game
            self.game = game.lower()
            self.wishlist = [self.game]
        self._state = None

    @property
    def entity_id(self):
        """Return the entity id of the sensor."""
        if self.game:
            return "sensor.nintendo_wishlist_{}".format(slugify(self.game))
        return "sensor.nintendo_wishlist"

    @property
    def name(self):
        """Return the name of the sensor."""
        if self._game:
            return self._game
        return "Nintendo Wishlist"

    @property
    def unit_of_measurement(self):
        """Return the unit of measurement of this entity, if any."""
        return "on sale"

    @property
    def state(self):
        """Return the state of the sensor."""
        return self._state

    @property
    def icon(self):
        """Icon to use in the frontend."""
        return "mdi:nintendo-switch"

    @property
    def device_state_attributes(self):
        return self.attrs

    async def async_update(self):
        """Get the latest data and updates the state."""
        if self.country in NA_COUNTRIES:
            await self._na_update()
        else:
            await self._eu_update()

    def _parse_matches(self, results: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        matches: List[Dict[str, Any]] = []
        for game in results:
            if not game["title"].lower().startswith(tuple(self.wishlist)):
                continue
            match = {
                "box_art_url": ("https://www.nintendo.com{}".format(game["boxArt"])),
                "normal_price": "${}".format(game["msrp"]),
                "percent_off": get_percent_off(game["msrp"], game["salePrice"]),
                "sale_price": "${}".format(game["salePrice"]),
                "title": game["title"],
            }
            matches.append(match)
        return matches

    async def _na_update(self):

        wish_list_matches = []
        queries = copy(QUERIES)
        params = queries[0]["params"]
        queries[0]["indexName"] = NA_INDEX_NAMES[self.country]
        async with SearchClient.create(APP_ID, API_KEY) as client:
            # Sets the default page to 0, if there are more pages we'll fetch those
            # after we know how many there are.
            query_params: str = params + "&page=0"
            queries[0]["params"] = query_params
            results = await client.multiple_queries_async(queries)
            wish_list_matches.extend(self._parse_matches(results["results"][0]["hits"]))
            # Check if there are more results than a single page.
            num_pages: int = results["results"][0]["nbPages"]
            if num_pages > 1:
                # Using pagination fetch the remaining results.
                for page_num in range(1, num_pages):
                    # The page param is a 0-based index.  Since we already fetched page
                    # 0, start at 1
                    query_params: str = params + "&page=" + str(page_num)
                    queries[0]["params"] = query_params
                    results = await client.multiple_queries_async(queries)
                    wish_list_matches.extend(
                        self._parse_matches(results["results"][0]["hits"])
                    )
        self.attrs["on_sale"] = wish_list_matches
        self._state = len(wish_list_matches)

    async def _eu_update(self):
        lang = COUNTRY_LANG[self.country]
        wish_list_matches = {}
        # NOTE: This endpoint requires the country to be lowercase.
        async with self.session.get(EU_SEARCH_URL.format(language=lang)) as resp:
            # The content-type is text/html so we need to specify None here.
            data = await resp.json(content_type=None)
            for game in data["response"]["docs"]:
                if not game["title"].lower().startswith(tuple(self.wishlist)):
                    continue
                wish_list_matches[game["nsuid_txt"][0]] = {
                    "box_art_url": "https:{}".format(game["image_url"]),
                    "percent_off": game["price_discount_percentage_f"],
                    "title": game["title"],
                }

                params = {
                    "country": self.country,
                    "ids": ",".join(wish_list_matches.keys()),
                    "lang": "en",
                }
                async with self.session.get(EU_PRICE_URL, params=params) as r:
                    prices = await r.json()
                    for price in prices["prices"]:
                        n_id = str(price["title_id"])
                        wish_list_matches[n_id]["normal_price"] = price[
                            "regular_price"
                        ]["amount"]
                        wish_list_matches[n_id]["sale_price"] = price["discount_price"][
                            "amount"
                        ]
        self.attrs["on_sale"] = list(wish_list_matches.values())
        self._state = len(wish_list_matches)
